helm-framework/helm/kubeConfig.go

```diff
@@ -0,0 +1,198 @@
+	// Use the appropriate types for the arguments
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryClient)
+	expander := restmapper.NewShortcutExpander(mapper, discoveryClient)
+
+	return expander, nil
+}
+
+// Function returning raw k8s client config
+func (k *KubeConfig) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	return k.ClientConfig
+}
+
+// Generates a k8s client config, based on providers settings and namespace, which this config will be used to interact with the k8s cluster
+func (m *Meta) newKubeConfig(namespace *string) (*KubeConfig, error) {
+	//Used to specify customer settings for k8s client config
+	overrides := &clientcmd.ConfigOverrides{}
+	//Used to load the k8s config files
+	loader := &clientcmd.ClientConfigLoadingRules{}
+	// Holds paths to k8s config files
+	configPaths := []string{}
+
+	//If config path is set, we will use it
+	if v := m.Data.Kubernetes.ConfigPath.ValueString(); v != "" {
+		configPaths = []string{v}
+		//If config paths is set, we append all paths to config paths
+	} else if !m.Data.Kubernetes.ConfigPaths.IsNull() {
+		for _, p := range m.Data.Kubernetes.ConfigPaths.Elements() {
+			configPaths = append(configPaths, p.(types.String).ValueString())
+		}
+		//if KUBE_CONFIG_PATHS is set, we split into indvidual paths
+	} else if v := os.Getenv("KUBE_CONFIG_PATHS"); v != "" {
+		configPaths = filepath.SplitList(v)
+	}
+	//If there are any config paths, we expand them to their full path
+	if len(configPaths) > 0 {
+		expandedPaths := []string{}
+		for _, p := range configPaths {
+			path, err := homedir.Expand(p)
+			if err != nil {
+				return nil, err
+			}
+			log.Printf("[DEBUG] Using kubeconfig: %s", path)
+			expandedPaths = append(expandedPaths, path)
+		}
+		// If there's only one path, we set it as the explicit path
+		if len(expandedPaths) == 1 {
+			loader.ExplicitPath = expandedPaths[0]
+			// If there is not, we set the precedence for the paths to be used by thr loaded
+		} else {
+			loader.Precedence = expandedPaths
+		}
+		// If ConfigContext is set, we overred the current context
+		if ctx := m.Data.Kubernetes.ConfigContext.ValueString(); ctx != "" {
+			overrides.CurrentContext = ctx
+			log.Printf("[DEBUG] Using custom current context: %q", overrides.CurrentContext)
+		}
+		// If ConfigContextAuthInfo is set, we override the auth info
+		if authInfo := m.Data.Kubernetes.ConfigContextAuthInfo.ValueString(); authInfo != "" {
+			overrides.Context.AuthInfo = authInfo
+		}
+		//If ConfigContextCluster is set, we override the cluster
+		if cluster := m.Data.Kubernetes.ConfigContextCluster.ValueString(); cluster != "" {
+			overrides.Context.Cluster = cluster
+		}
+		log.Printf("[DEBUG] Using overridden context: %#v", overrides.Context)
+	}
+	//Checking whether or not to override the tls releated settings
+	if v := m.Data.Kubernetes.Insecure.ValueBool(); v {
+		overrides.ClusterInfo.InsecureSkipTLSVerify = v
+	}
+	if v := m.Data.Kubernetes.TlsServerName.ValueString(); v != "" {
```

Comment:
I think most of these `v != ""` checks are superfluous now. The field in the struct is a `string` which defaults to `""` if you don't set anything. So it doesn't matter here if `ValueString()` returns `""`.